### PR TITLE
Add IVSPathContextProvider method to get only the user wide available configuration

### LINKED ISSUE #9395: Add interop API to enumerate fallback folders without requiring a solution/project
## Environment Details
NuGet product used: VS UI
NuGet version: 5.6.0.6489
VS version: 16.7
Worked before?: No

## Feature Request
NuGet settings can be set in the following locations:
- machine-wide (%ProgramFiles(x86)%\NuGet\Config\\*.config)
- user-specific (%APPDATA%\NuGet\NuGet.config)
- solution-wide

The XAML and WinForms teams need an API to enumerate machine-wide and user-specific fallback folders that:
1. Works without requiring a solution to be open
2. Ignores solution-wide settings if a solution is loaded

In order to call this API from within VS, it needs to be an interop API.

Likely choices to support from NuGet according to @rrelyea:
1. Tweak existing VS interop API to allow null value (or similar) to get what you need. Note that the requirement to ignore solution config when a solution is open may make this option impossible.
2. Add a new VS interop API.

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9395
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 

Add a new API to provide a path context from only the user wide available configuration.
The motivation behind this is covered in https://github.com/NuGet/Home/issues/9395, and it'll be used by the xaml team to discover controls. 

If a solution is loaded, note that the values in the path context might not be the actual effective values for the solution.
If a customer has overriden the `globalPackagesFolder` key or cleared the `fallbackPackageFolders`, these values will be incorrect.
It is important to keep this scenario in mind when working with this path. To predict differences you can call this in combination with <see cref="TryCreateSolutionContext(out IVsPathContext2)"/>.

@debonte is aware of this. 

Originally I wanted this method to be async, but I forgot that COM doesn't have support for generics like Task<T>, so had to make it sync.  

Finally note that the results can be safely cached because they can't really be edited.

Docs issue: https://github.com/NuGet/docs.microsoft.com-nuget/issues/2084

Lastly, this fix is needed for 5.7 :) 

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:  Manual with https://github.com/NuGet/Entropy/blob/master/IVsTestingExtension/README.md to test for deadlocks + tests.

### CODE REVIEW

**Review by srdjanjovcic — DISMISSED** [review]
:shipit:

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs [hunk 1]
```diff
@@ -318,5 +327,24 @@ private async Task<IEnumerable<Project>> GetProjectsInSolutionAsync(DTE dte)
             IEnumerable<Project> supportedProjects = allProjects.Where(EnvDTEProjectUtility.IsSupported);
             return supportedProjects;
         }
+
+        public bool TryCreateUserWideContext(out IVsPathContext vsPathContext)
+        {
+            // invoke async operation from within synchronous method
+            vsPathContext = NuGetUIThreadHelper.JoinableTaskFactory.Run(() => TryCreateUserWideContextAsync());
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs]: Would it be worthwhile throwing if we're on the UI thread, to make sure callers call us on a background thread?

Although if a C++ component is calling us through COM, I don't know if that's possible.

File: src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs [hunk 2]
```diff
@@ -42,5 +44,20 @@ public interface IVsPathContextProvider2 : IVsPathContextProvider
...
+        /// If a customer has overriden the `globalPackagesFolder` key or cleared the `fallbackPackageFolders`, these values will be incorrect.
+        /// It is important to keep this scenario in mind when working with this path. To predict differences you can call this in combination with <see cref="TryCreateSolutionContext(out IVsPathContext2)"/>.
+        /// </summary>
+        /// <returns>
+        /// <code>True</code> if operation has succeeded and context was created.
+        /// <code>False</code> otherwise.
+        /// </returns>
+        /// <throws>
+        /// <code>InvalidOperationException</code> when it fails to create a context and return appropriate error message.
+        /// </throws>
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: There is no `<throws>` xml doc element documented, but there is an [`<exception>`](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/exception). Did you hover the mouse over the method name to see if IntelliSense shows the docs?

**Review by zivkan — APPROVED** [review]

**Comment by srdjanjovcic** [comment]
>     public interface IVsPathContextProvider2 : IVsPathContextProvider

Where is this interface originally defined? COM interfaces are immutable once published, so you cannot add members this way

---
Refers to: src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs:15 in 8b9856f. [](commit_id = 8b9856f4d32021f9116097b2ef0c0c1f091b9b87, deletion_comment = False)

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs (hunk 1 above)
Comment (srdjanjovcic) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs]: You can always explicitly switch to background thread:

await TaskScheduler.Default;

---
In reply to: [446392155](https://github.com/NuGet/NuGet.Client/pull/3478#discussion_r446392155) [](ancestors = 446392155)

File: src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs [hunk 3]
```diff
@@ -42,5 +44,20 @@ public interface IVsPathContextProvider2 : IVsPathContextProvider
...
+        bool TryCreateUserWideContext(out IVsPathContext vsPathContext);
```
Comment (debonte) [inline src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: When I first saw the name TryCreateUserWideContext, I was thinking that "UserWide" meant that it only contained the user-specific settings and not the machine-wide settings. Is this a term that consumers will understand because it's used in other NuGet APIs?

If not, maybe it needs a better name? I don't have any spectacular ideas, but maybe something like TryCreateSolutionlessContext or TryCreateNoSolutionContext?

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: It's something used in our docs.
I can't with high confidence claim that it'll obvious to customers. 

Others thoughts?

**Comment by nkolev92** [comment]
> Where is this interface originally defined? COM interfaces are immutable once published, so you cannot add members this way

@srdjanjovcic 

I was under the impression that defining additional methods is possible as long as they are ordered. 
Would you recommending adding IVSPathContextProvider3 instead? 
I'd be happy to do that if necessary.

File: src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: I'll try it out.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs]: I'll add a comment that this API can be safely invoked on a background thread. 
The threading logic is similar as the other IVSPathContextProvider methods.

**Comment by srdjanjovcic** [comment]
Why is this COM in the first place?

---
In reply to: [650407430](https://github.com/NuGet/NuGet.Client/pull/3478#issuecomment-650407430) [](ancestors = 650407430)

**Comment by nkolev92** [comment]
> Why is this COM in the first place?

Embedding purposes. 
We've mostly been adding methods, but not changing *all* of our APIs. 
https://github.com/NuGet/NuGet.Client/pull/3469 contains the new extensibility solution, but in the interest of getting this in 16.7, we're continuing with what we have.

**Comment by srdjanjovcic** [comment]
@nkolev92 Adding member to COM interface is considered a breaking change because COM has to account for all combinations of usage, when interface has same GUID. It would break new client talking to old server.

Now, this particular usage has no such problems because, apparently, this is bending the COM to be used in tightly coupled client-server scenario. NuGet client VSIX is always shipped as a payload of Visual Studio, right? If so, interface and client usage always match. Problem would appear if NuGet was true extension, so that users can install it in whatever version of VS they want, which would conceivably lead to mismatched interface implementation and interface usage (new VSIX installed in [still supported] 16.0)

**Comment by nkolev92** [comment]
Thanks, that makes sense. 
Yeah we are unlikely to have `new client` <=> `old server` scenarios. 
We *will* have `old client` <=> `new server` scenario, but we are not removing methods. 

I don't have a preference over adding a new interface vs adding a new method, so I'd be happy to change if you think so. (It's super quick to do so :) )

**Comment by srdjanjovcic** [comment]
Re: COM interfaces, there is no COM police to enforce COM rules 🙂, so, as long as you do not call old server from new client, you are good.

Yo can even change GUID, if you want to enforce change. Then, you have two different interfaces which just happened to be named same.

In any case, I am fine with extending it as-is, particularly, if we are going away from COM interfaces and will expose functionality through proper VS service.

**Review by srdjanjovcic — APPROVED** [review]

File: src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs (hunk 3 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: I'd vote for something with "no solution" in the name. I like `TryCreateNoSolutionContext`.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: done.

File: src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: fixed.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs]: I'd like to leave this as is. 

> Would it be worthwhile throwing if we're on the UI thread, to make sure callers call us on a background thread?

This seems like an anti-pattern to me :) 

I documented the threading affinity.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs [hunk 4]
```diff
@@ -318,5 +327,24 @@ private async Task<IEnumerable<Project>> GetProjectsInSolutionAsync(DTE dte)
...
             return supportedProjects;
         }
+
+        public bool TryCreateNoSolutionContext(out IVsPathContext vsPathContext)
+        {
+            // invoke async operation from within synchronous method
+            vsPathContext = NuGetUIThreadHelper.JoinableTaskFactory.Run(() => TryCreateUserWideContextAsync());
+
+            return vsPathContext != null;
+
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs]: nit: unnecessary whitespace line

File: src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs [hunk 5]
```diff
@@ -42,5 +44,21 @@ public interface IVsPathContextProvider2 : IVsPathContextProvider
...
+        /// Attempts to create an instance of <see cref="IVsPathContext"/> containing only the user wide and machine wide configurations.
+        /// If a solution is loaded, note that the values in the path context might not be the actual effective values for the solution.
+        /// If a customer has overriden the `globalPackagesFolder` key or cleared the `fallbackPackageFolders`, these values will be incorrect.
+        /// It is important to keep this scenario in mind when working with this path. To predict differences you can call this in combination with <see cref="TryCreateSolutionContext(out IVsPathContext2)"/>.
+        /// </summary>
+        /// <returns>
+        /// <code>True</code> if operation has succeeded and context was created.
+        /// <code>False</code> otherwise.
+        /// </returns>
+        /// <exception cref="InvalidOperationException">When we fail to create a context and return appropriate error message</exception>
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: If we throw an exception when we can't create the context, when will `false` ever be returned?  Given this is a `Try` method, we should never throw, only return false, or we should return the context, not a bool, and remove `Try` from the name?

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPathContextProvider2.cs]: We will never throw. 
Deleting it. 
I am trying to keep the same calling pattern as the rest of the APIs.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs]: done.
